peps/pep-0748.rst

```diff
@@ -0,0 +1,923 @@
+support. There are many security advantages in refusing to use old versions of
+TLS, and some misbehaving servers will mishandle TLS clients advertising support
+for newer versions.
+
+The following enumerated type can be used to gate TLS versions. Forward-looking
+applications should almost never set a maximum TLS version unless they
+absolutely must, as a TLS backend that is newer than the Python that uses it may
+support TLS versions that are not in this enumerated type.
+
+Additionally, this enumerated type defines two additional flags that can always
+be used to request either the lowest or highest TLS version supported by an
+implementation. As for cipher suites, we expect that for most users, secure
+defaults will be enough. When specifying no list of TLS versions, the backends
+should use secure defaults (possibly derived from system recommended settings).
+
+.. code-block:: python
+
+    TODO TLSVersion from tlslib
+
+Errors
+~~~~~~
+
+This module would define four base classes for use with error handling. Unlike
+many of the other classes defined here, these classes are not abstract, as they
+have no behavior. They exist simply to signal certain common behaviors. Backends
+should subclass these exceptions in their own packages, but needn't define any
+behavior for them.
+
+In general, concrete implementations should subclass these exceptions rather
+than throw them directly. This makes it moderately easier to determine which
+concrete TLS implementation is in use during debugging of unexpected errors.
+However, this is not mandatory.
+
+The definitions of the errors are below:
+
+.. code-block:: python
+
+    TODO errors from tlslib
+
+Certificates
+~~~~~~~~~~~~
+
+This module would define a concrete certificate class. This class would have
+almost no behavior, as the goal of this module is not to provide all possible
+relevant cryptographic functionality that could be provided by X.509
+certificates. Instead, all we need is the ability to signal the source of a
+certificate to a concrete implementation.
+
+For that reason, this certificate class defines three attributes, corresponding
+to the three envisioned constructors: certificates from files, certificates from
+memory, or certificates from arbitrary identifiers. It is possible that backends
+do not support all of these constructors, and they can communicate this to users
+as described in the “Runtime” section below.
+
+Specifically, this class does not parse any provided input to validate that it
+is a correct certificate, and also does not provide any form of introspection
+into a particular certificate. Backends are not required to provide such
+introspection either. Peer certificates that are received during the handshake
+are provided as raw DER bytes.
+
+Future versions of the API may provide alternative constructors, e.g. to load
+certificates from HSMs, if a common interface emerges for doing this.
```

Comment:
I suspect the "ID" interface will be enough to support HSMs (they generally have flat lists of keys and certs and refer to them by index)

Other readers may not know that though, so may get the impression this design wouldn't be usable with HSMs/TPMs/SSMs yet